Simplified/Reader2/NYPLReaderBookmarksBusinessLogic.swift

```diff
@@ -337,7 +344,99 @@ class NYPLReaderBookmarksBusinessLogic: NSObject, NYPLReadiumViewSyncManagerDele
     completion()
   }
 
-  // TODO: Sync reading position
+  // MARK: - Read Position
+  
+  func storeReadPosition(locator: Locator) {
+    // Avoid overwriting location when reader first open
+    guard (locator.locations.totalProgression ?? 0) != 0,
+      let bookLocation = NYPLBookLocation(locator: locator, publication: publication, renderer: NYPLBookLocation.r2Renderer) else {
+      return
+    }
+    
+    bookRegistryProvider.setLocation(bookLocation, forIdentifier: book.identifier)
+    postReadPosition(locationString: bookLocation.locationString)
+  }
+  
+  /// Restore locally stored location and  sync location from server
+  /// Both local and server fetch completions are only called if fetch succeed
+  func restoreReadPosition(currentLocator: Locator?, localFetchCompletion: (Locator?) -> (), serverFetchCompletion: @escaping (Locator?) -> ()) {
+    var currentLocation:NYPLBookLocation?
+    
+    if let currentLocator = currentLocator {
+      currentLocation = NYPLBookLocation(locator: currentLocator, publication: publication, renderer: NYPLBookLocation.r2Renderer)
+    }
+    
+    if let lastSavedLocation = bookRegistryProvider.location(forIdentifier: book.identifier),
+      let locator = lastSavedLocation.convertToLocator() {
+      localFetchCompletion(locator)
+      currentLocation = lastSavedLocation
+    }
+    
+    syncReadPosition(bookID: book.identifier, currentLocation: currentLocation, url: book.annotationsURL) { (serverLocationString) in
+      guard let locationString = serverLocationString,
+        let locator = NYPLBookLocation(locationString: locationString, renderer: NYPLBookLocation.r2Renderer).convertToLocator() else {
+        Log.debug(#file, "Unable to create locator from server stored location string - \(serverLocationString ?? "")")
+        return
+      }
+      
+      serverFetchCompletion(locator)
+    }
+  }
+  
+  /// Post the read position to server if we have synced the read position from server
+  /// There is a 120 seconds interval to avoid high frequency of requests
+  /// - Parameters:
+  /// - locationString: A json string that contains required information to create a Locator object
+  private func postReadPosition(locationString: String) {
+    serialQueue.async {
+      self.queuedReadPosition = locationString
+      
+      if Date() > self.lastReadPositionUploadDate.addingTimeInterval(120) {
+        self.postQueuedReadPosition()
+      }
+    }
+  }
+  
+  @objc private func postQueuedReadPosition() {
```

Comment:
this function seems to be called from 2 contexts:
- from inside the `serialQueue` (i.e. from `postReadPosition()`)
- from outside the `serialQueue` (i.e. as a NotificationCenter observer callback)

in the latter case we are accessing shared instance data outside the `serialQueue`, so we have race conditions. 

One way to fix this would be to add a new function that calls `postQueuedReadPosition()` from inside the `serialQueue`, and then update the NotificationCenter observation to call that new function instead of `postQueuedReadPosition`.